### ISSUE TITLE
[FR] --target minor or patch should exclude 0.x.x release

### ISSUE THREAD

**Issue by ashidaharo** [raineorshine/npm-check-updates#958]
- [x] I have searched for [similar issues](https://github.com/raineorshine/npm-check-updates/issues)
- [x] I am using the latest version of `npm-check-updates`
- [x] I am using `node >= 10.17`

---------------------------

## Steps to Reproduce
```
ncu --target minor
```
or
```
ncu --target patch
```
## Current Behavior
Both results include the new release of 0.x.x.
like
```
ncu --target minor
 axios                              ^0.22.0  →   ^0.23.0
```
## Expected Behavior
0.x.x means that all release MAY contain BREAKING CHANGES.
So, release of 0.x.x should be  treated as a MAJOR release.
This should be a feasible feature, since it is correctly displayed as such when --color is set.

**Comment by ashidaharo** [comment]
But... this fix seems to me to be a breaking change.
I can't decide whether to include it in the patch version as a bug fix or in the next major version update as a breaking change, so I guess I can't send you a PR for it...

**Comment by ashidaharo** [comment]
> @ashidaharo I've just discovered it too) It should be something that was left intentionally or a new bug (but I think it was always like this). I should not think nobody hasn't noticed it during all this time, for it's a very important behavior. But what was the intension?

When I found the source of the problem, I confirmed that this was a behavior that had always existed, whether it was intentional or not. I think it's probably just that the person who originally wrote it didn't pay attention to exceptions like '0.x.x'.

However, the current behavior has been confirmed as a specification by the following tests.
https://github.com/raineorshine/npm-check-updates/blob/26ceb1e5aa81592adbd0d45a68ad4c7cbfbb7a62/test/version-util.test.js#L393-L396
For that reason, I considered the current behavior of this library to be as per the specification of this package, although it does not follow the SemVer specification. Therefore, I thought that fixing this problem should be treated as a destructive change, not a bug.

**Comment by lunany188** [comment]
I've just noticed that the version has been upgraded, so I came to remind everyone.
The problem is still unresolved in any way, and there continues to be a difference in the actual behavior of coloring and `--target`.

**Comment by raineorshine** [comment]
The coloring uses semver, while `target` refers simply to the digit (whatever that's worth). 

I imagine that `--target semver` is what a lot of people want (#581 for historical discussion).

**Comment by kokushkin** [comment]
> target refers simply to the digit (whatever that's worth)

but this is how semver also treats it. Minor in semver is just the second digit!

**Comment by raineorshine** [comment]
> > target refers simply to the digit (whatever that's worth)
> 
> but this is how semver also treats it. Minor in semver is just the second digit!

In that context yes, but the ambiguity arises with a leading `0.`. The version `0.1.0` has a minor digit of `1`, but changing that minor digit results in a major upgrade, NOT a minor upgrade. Thus it depends on the context in which we use the word "minor".

TLDR; "minor" digit is different than "minor" upgrade.

**Comment by ljharb** [comment]
indeed, it’s only >= v1 that minor is always the second digit.

**Comment by kokushkin** [comment]
> TLDR; "minor" digit is different than "minor" upgrade.

yeah, here we go) "minor" update doesn't exist in semver though, and "minor" version very often confused with "^"

**Comment by ljharb** [comment]
a "minor" upgrade in semver does exist, because semver describes how you can determine breakage between two semver versions. `^` is a semver range defined mostly by npm, which isn't in the spec yet but will likely be soon.

**Comment by kokushkin** [comment]
Not entirely true I'd say. "^" doesn't mean exactly no breaking changes (according to semver)

**Comment by ljharb** [comment]
@kokushkin semver doesn't say anything about `^`. In npm, and the open PR adding ranges to the semver spec, it EXPLICITLY says "no breaking changes".

**Comment by kokushkin** [comment]
I didn't mean "^" defined in semver. I know nothing about this PR. If it is accepted then it will be a new version of semver I suppose.

**Comment by ashidaharo** [comment]
I would like to offer a opinion from a different perspective from the current discussion.
> I imagine that --target semver is what a lot of people want (#581 for historical discussion).

It would be nice to see such a feature implemented, but apart from that. 
**It's not intuitive that the color coding and the behavior of `target` are different, and I'd like to see at least one of them unified.**
Of course, I understand that breaking changes to the behavior of `target` are undesirable, and the same is true for the color coding displayed by digits, but this is the main motivation for creating this issue!

**Comment by raineorshine** [comment]
Fair enough. From my perspective, coloring should always be semver for simplicity (not varied based on target), and `--target semver` should be added instead of changing `--target minor`. At time of writing there is only one "+1" on this issue, so I don't see any sizable support to make a breaking change.

**Comment by ashidaharo** [comment]
Understood. If a new feature is provided that works the same way as coloring, no matter how different it is from this original issue, then that is what I really wanted.

**Comment by kokushkin** [comment]
I think I've got what had confused me at the beginning. It was the blue color.
![image](https://user-images.githubusercontent.com/9853656/149980528-14a9e44c-753d-4ab8-b4d8-52e4108a827b.png)
If everything was green (not breaking) or read (breaking) and  if major.minor.patch are just 1st 2nd 3rd then it's all fine for me, I don't need the blue color.

Otherwise, blue might be associated with minor, then red associated with major.

But in this case
![image](https://user-images.githubusercontent.com/9853656/149983188-c9807587-1344-4553-998f-6f3f8c6da38c.png)
It's not a major, neither from my understanding of it as just the 3rd number (should be green then), nor from other's understanding of it as a minor of 0.x (should be blue then).

Therefore, my suggestion is to screw the blue color entirely)

**Comment by ashidaharo** [comment]
In my opinion, [semver definition](https://semver.org/#spec-item-4) says "Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable." So, even if the 0.0.x update is treated as a non-breaking update by convention, i think it should be treated as a breaking change by definition, since it is only a convention.
Of course, this is only the definition of the original semver, so if [the npm documentation](https://github.com/npm/node-semver/blob/main/README.md) clearly states that it must be handled as such, that's a different story.
However, it only says this, "However, it presumes that there will not be breaking changes between 0.2.4 and 0.2.5. It allows for changes that are presumed to be additive (but non-breaking), according to **commonly observed practices**".

**Comment by ljharb** [comment]
@ashidaharo That's v2 of the semver spec. npm follows v1 of the spec, and "what npm does" is the arbiter of truth for the npm ecosystem. npm's rules are https://github.com/raineorshine/npm-check-updates/issues/958#issuecomment-952323019 - there's no room for debate.

**Comment by ashidaharo** [comment]
@ljharb 
1. This definition has not changed since [semver v1](https://semver.org/spec/v1.0.0.html#spec-item-6).
2. [npm document](https://github.com/npm/node-semver/blob/main/README.md#versions) says that it follows semver v2.

**Comment by ljharb** [comment]
none the less, semver.npmjs.com and the semver package treats it the way i've described, so all that means is the docs are wrong.

**Comment by ashidaharo** [comment]
> none the less, semver.npmjs.com and the semver package treats it the way i've described, so all that means is the docs are wrong.

Yes, [that is mentioned](https://github.com/npm/node-semver/blob/main/README.md#caret-ranges-123-025-004).
> Many authors treat a 0.x version as if the x were the major "breaking-change" indicator.

However, this is not a definition, but an implementation of npm's Semver.
As I quoted earlier, the actual npm packages also mention that this is not always expected, even though some people certainly treat it as a common practice.

**Comment by ljharb** [comment]
The implementation is the only thing that matters, since that's how `^` actually behaves and has always behaved in the npm ecosystem.

**Comment by ashidaharo** [comment]
It looks like we will have to wait for the response of the maintainer and others who are interested in this issue, as we cannot seem to reach a mutual agreement on this.

**Comment by ljharb** [comment]
I mean sure, but what would the point be of "npm check updates" if it wasn't checking "what npm will actually update"

**Comment by ashidaharo** [comment]
> npm-check-updates upgrades your package.json dependencies to the latest versions, ignoring specified versions.

https://github.com/raineorshine/npm-check-updates#npm-check-updates

**Comment by ashidaharo** [comment]
If you want to update safely according to npm's behavior, use ``npm update`` instead of this tool, which is sufficient for minor updates, and if you want to do major updates, you should check with ``npm outdated`` and then update manually.

**Comment by ashidaharo** [comment]
And in that way, you can use not only Caret but also Tilde, X-Ranges, and various other version control methods provided by npm.

**Comment by ashidaharo** [comment]
From the beginning, this tool was a tool that worked independently of the npm implementation, or rather, actively ignored it, even before the target option was added.

**Comment by raineorshine** [comment]
> It's not a major, neither from my understanding of it as just the 3rd number (should be green then), nor from other's understanding of it as a minor of 0.x (should be blue then).
> 
> Therefore, my suggestion is to screw the blue color entirely)

I take your point. It's not a binary breaking-nonbreaking distinction, and it's not strictly 1st 2nd 3rd, although it's close to that. As described in the README:

- Red = major upgrade (and all major version zero)
- Cyan = minor upgrade
- Green = patch upgrade

I actually think this is still the best representation. Showing `0.x` upgrades in green is dangerous, and defeats the purpose of red as a distinctive color for breaking changes. Blue could be removed I suppose, but that seems like an unnecessary loss of functionality. Blue adds an extra layer of distinction that can be ignored by most at no cost. Though I see your point that it may not be no cost for those who are confused by the blue color.

> > npm-check-updates upgrades your package.json dependencies to the latest versions, ignoring specified versions.
> 
> [raineorshine/npm-check-updates#npm-check-updates](https://github.com/raineorshine/npm-check-updates#npm-check-updates)

That's right. npm-check-updates has primarily existed to forcefully break versions, since this functionality is not in npm. Over time, developers have come to use npm-check-updates as an alternative to `npm outdated` and for more fine-grained control of package upgrades. There's a real need for `--target semver`, we just need a contributor.

**Comment by kokushkin** [comment]
so --target semver would show what?
let's say I have 1.2.3 dependency  (not ^1.2.3 or ~1.2.3) , and there're 1.2.6, 1.3.0, 2.0.1 , what should --target semver show me?

**Comment by raineorshine** [comment]
> so --target semver would show what?
> let's say I have 1.2.3 dependency (not ^1.2.3 or ~1.2.3) , and there're 1.2.6, 1.3.0, 2.0.1 , what should --target semver show me?

Well, I suppose I would want `--target semver` to match `npm update`. Then I would want `--target caret` to upgrade all dependencies as if they started with `^`.

**Comment by kokushkin** [comment]
I'm not sure if npm update does anything about dependencies without  ^ or ~ (rather nothing). --target semver also doesn't reflect what is written in semver in case of 0.x.y.  Would it be better to have just --target caret  (updates everything as if it has ^), and --target tilde (updates everything as if it has ~)?
I've just recalled me adding ^ before every dependency and calling npm update after ;)
But why it shouldn't be added in npm update itself then?

**Comment by raineorshine** [comment]
> Would it be better to have just --target caret (updates everything as if it has ^), and --target tilde (updates everything as if it has ~)?

Yes, I like that!

**Comment by Primajin** [comment]
I am currently using the `--filterVersion` option in order to run the checks only on non-pinned versions.
However I need to run it twice since
```
> npm-check-updates --filterVersion "/^[~^]/" -t minor

Checking C:\mypath\package.json
[====================] 130/130 100%

 @types/node                       ^17.0.9  →  ^17.0.10
 @types/react-router               ^5.1.17  →   ^5.1.18
 @typescript-eslint/eslint-plugin   ~5.9.0  →   ~5.10.0
 @typescript-eslint/parser          ~5.9.0  →   ~5.10.0
 eslint-plugin-jest                ~25.3.4  →   ~25.7.0
 npm-check-updates                 ^12.1.0  →   ^12.2.0
 typescript                         ^4.5.4  →    ^4.5.5
```
gives me not exactly want I want.
I actually don't want to bump a `minor` (`^`) on @typescript-eslint/parser but only a `patch` (`~`)

So instead I changed that to
```
> npm-check-updates --filterVersion "/^[\^]/" -t minor && npm-check-updates --filterVersion "/^[~]/" -t patch

Checking C:\mypath\package.json
[====================] 127/127 100%

 @types/node          ^17.0.9  →  ^17.0.10
 @types/react-router  ^5.1.17  →   ^5.1.18
 npm-check-updates    ^12.1.0  →   ^12.2.0
 typescript            ^4.5.4  →    ^4.5.5

Run ncu -u to upgrade package.json

Checking C:\maypath\package.json
[====================] 3/3 100%

 @typescript-eslint/eslint-plugin  ~5.9.0  →  ~5.9.1
 @typescript-eslint/parser         ~5.9.0  →  ~5.9.1

Run ncu -u to upgrade package.json
```

Which gives me the "correct" minor and patches for the ones I want and skips pinned versions.

Though I wonder if we could introduce a new target version like `safe` (naming is hard, maybe you have a better one) that does that automatically:

1. Skips all pinned versions
2. Checks for minor on caret
3. Checks for patch on tilde
4. **maybe even skips versions that start with 0**

What do you think?

**Comment by raineorshine** [comment]
@Primajin Please open a new issue and I will respond there. This issue is specifically about whether `--target minor` should exclude 0.x.x releases.

**Comment by Primajin** [comment]
Sure happy to - just thought two birds one stone 😅

**Comment by raineorshine** [comment]
As of `v12.5.0`, you can specify a custom target function in your .ncurc.js file, or when importing npm-check-updates as a module:

```js
/** Custom target.
  @param dependencyName The name of the dependency.
  @param parsedVersion A parsed Semver object from semver-utils.
    (See https://git.coolaj86.com/coolaj86/semver-utils.js#semverutils-parse-semverstring)
  @returns One of the valid target values (specified in the table above).
*/
target: (dependencyName, [{ semver, version, operator, major, minor, patch, release, build }]) => {
  if (major === 0) return 'minor'
  return 'latest'
}
```

You can hack together your own semver-like target logic, but the intention is to eventually add a `--semver` option that handles this for you.

**Comment by betaorbust** [comment]
As a note: on my machine the major tag came through as a string, instead of a number.
```diff
-if (major === 0) return 'minor'
+if (major === '0') return 'minor'
```

@raineorshine Thanks for your time and effort making such a great tool!

**Comment by Primajin** [comment]
@betaorbust You can also just use two equals `==` then it doesn't matter if it's a string or a number :wink:

**Comment by betaorbust** [comment]
_Chaotic Alignment has entered the chat_

**Comment by raineorshine** [comment]
Closing, as described in https://github.com/raineorshine/npm-check-updates/issues/958#issuecomment-1017594507.

For the proposed `--semver` option, see #1054.
